### PR TITLE
Chore: Refactor Open With based on API changes

### PR DESCRIPTION
**Pull request by jeremypress**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @jeremypress has signed the CLA. Thanks for the pull request!

**Comment by boxcla** [comment]
# Webpack Bundle Size Github Bot

| Bundle        | Diff | Percentage Diff | Before | After |
| ------------- |:----:|:---------------:|:------:|:-----:|
| Total | 0 B | 0.00% | 0 B | 0 B |

Compared against previously merged PR #540

**Powered by [webpack-bundle-size-github-bot](https://github.com/wenboyu2/webpack-bundle-size-github-bot/)**

File: src/components/OpenWith/OpenWith.js [hunk 1]
```diff
@@ -112,7 +121,26 @@ class OpenWith extends PureComponent<Props, State> {
...
+     * @return {void}
+     */
+    fetchOpenWithData() {
+        const { fileId, extension }: Props = this.props;
+        if (extension) {
+            this.setState({ extension });
+        } else {
+            this.api
+                .getFileAPI()
+                .getFile(fileId, this.fetchExtensionSuccessCallback, noop, {
```
Comment (ConradJChan) [inline src/components/OpenWith/OpenWith.js]: There's no dependency between this call (when there's no extension) with the `getOpenWithIntegrations` call right?

Comment (jeremypress) [reply src/components/OpenWith/OpenWith.js]: Correct, the icon can async load before or after we get back integrations. We'll use a generic icon until this comes back.

File: src/components/OpenWith/OpenWith.js [hunk 2]
```diff
@@ -112,16 +121,60 @@ class OpenWith extends PureComponent<Props, State> {
...
+    }
+
+    /**
+     * Fetches file extension and Open With data.
+     *
+     * @return {void}
+     */
+    fetchOpenWithData() {
+        const { fileId, extension, language }: Props = this.props;
+        if (extension) {
```
Comment (jstoffan) [inline src/components/OpenWith/OpenWith.js]: It seems like this operation should happen separately: once in the `initialState` construction and once again in componentDidUpdate if the extension changed.

File: src/components/OpenWith/OpenWith.js [hunk 3]
```diff
@@ -112,16 +121,60 @@ class OpenWith extends PureComponent<Props, State> {
...
+    /**
+     * After component updates, re-fetch Open With data if appropriate.
+     *
+     * @return {void}
+     */
+    componentDidUpdate(prevProps: Props): void {
+        const { fileId: currentFileId }: Props = this.props;
+        const { fileId: previousFileId }: Props = prevProps;
+
+        if (currentFileId !== previousFileId) {
```
Comment (jstoffan) [inline src/components/OpenWith/OpenWith.js]: Would we only want to do this if the extension changes?

File: src/components/OpenWith/OpenWith.js [hunk 4]
```diff
@@ -112,16 +121,60 @@ class OpenWith extends PureComponent<Props, State> {
...
-        const { fileId }: Props = this.props;
+        this.fetchOpenWithData();
+    }
+
+    /**
+     * Fetches file extension and Open With data.
+     *
+     * @return {void}
+     */
+    fetchOpenWithData() {
```
Comment (jstoffan) [inline src/components/OpenWith/OpenWith.js]: Should this be named `fetchFileExtension`?

File: src/api/OpenWith.js [hunk 5]
```diff
@@ -5,14 +5,11 @@
  */
 
 import Base from './Base';
-import AppIntegrationsAPI from './AppIntegrations';
+// Temporary until API is fixed
```
[the remark this answers is not in the document]
Comment (jeremypress) [reply src/api/OpenWith.js]: yep, forgot to update.

File: src/components/OpenWith/OpenWith.js (hunk 4 above)
Comment (jeremypress) [reply src/components/OpenWith/OpenWith.js]: It currently fetches the extension and the open with data

File: src/components/OpenWith/OpenWith.js (hunk 3 above)
Comment (jeremypress) [reply src/components/OpenWith/OpenWith.js]: If the file ID changes, then we need to get new open with data and a new extension. Either that will be provided via props or we'll fetch for it.

File: src/components/OpenWith/OpenWith.js (hunk 2 above)
Comment (jeremypress) [reply src/components/OpenWith/OpenWith.js]: The extension can/should only change if the file ID changes. That's why I have them grouped currently.

File: src/components/OpenWith/OpenWith.js (hunk 3 above)
Comment (jstoffan) [reply src/components/OpenWith/OpenWith.js]: Ah, okay. I didn't realize the method was doing two different things. It seems like it should still be two methods to me, but it's not really a big deal.

**Pull request merged**